File: src/test/java/com/example/librarymanagement/service/impl/BookPresenceImplTest.java
package com.example.librarymanagement.service.impl;

import com.example.librarymanagement.data.TestDataFactory;
import com.example.librarymanagement.exception.BookNotAvailableException;
import com.example.librarymanagement.exception.EntityNotFoundException;
import com.example.librarymanagement.model.entity.Book;
import com.example.librarymanagement.model.entity.BookPresence;
import com.example.librarymanagement.model.entity.Journal;
import com.example.librarymanagement.model.entity.Library;
import com.example.librarymanagement.model.entity.Reservation;
import com.example.librarymanagement.model.entity.User;
import com.example.librarymanagement.model.enums.Availability;
import com.example.librarymanagement.repository.BookPresenceRepository;
import com.example.librarymanagement.repository.BookRepository;
import com.example.librarymanagement.repository.LibraryRepository;
import com.example.librarymanagement.repository.ReservationRepository;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class BookPresenceImplTest {
    @InjectMocks
    private BookPresenceServiceImpl bookPresenceServiceImpl;

    @Mock
    private BookPresenceRepository bookPresenceRepository;

    @Mock
    private BookRepository bookRepository;

    @Mock
    private LibraryRepository libraryRepository;

    @Mock
    private ReservationRepository reservationRepository;

    @Mock
    private JournalServiceImpl journalService;

    private static BookPresence bookPresence;
    private static Library library;
    private static Book book;
    private static User user;
    private static Journal journal;
    private static Reservation reservation;

    @BeforeAll
    public static void init() {
        TestDataFactory.TestDataRel data = TestDataFactory.createTestDataRel();
        bookPresence = data.bookPresence;
        library = data.library;
        book = data.book;
        user = data.user;
        journal = data.journal;
        reservation = data.reservation;
    }

    @Test
    public void createBookPresence() {
        when(bookPresenceRepository.save(bookPresence)).thenReturn(bookPresence);

        assertEquals(bookPresence, bookPresenceServiceImpl.createBookPresence(bookPresence));
        verify(bookPresenceRepository, times(1)).save(bookPresence);
    }

    @Test
    public void addUserToBook() {
        when(bookPresenceRepository
                .findAllByLibraryIdAndBookIdAndAvailability(library.getId(), book.getId(), Availability.AVAILABLE))
                .thenReturn(List.of(bookPresence));
        when(bookPresenceRepository.save(any(BookPresence.class))).thenReturn(bookPresence);
        when(journalService.createJournal(any(Journal.class))).thenReturn(journal);

        assertEquals(bookPresence, bookPresenceServiceImpl.addUserToBook(user, library.getId(), book.getId()));
        verify(bookPresenceRepository, times(1))
                .findAllByLibraryIdAndBookIdAndAvailability(1L, 1L, Availability.AVAILABLE);
        verify(bookPresenceRepository, times(1)).save(any(BookPresence.class));
        verify(journalService, times(1)).createJournal(any(Journal.class));
    }

    @Test
    public void addUserToBook_BookNotAvailable() {
        when(bookPresenceRepository
                .findAllByLibraryIdAndBookIdAndAvailability(library.getId(), book.getId(), Availability.AVAILABLE))
                .thenReturn(List.of());

        assertThrows(BookNotAvailableException.class, () -> {
            bookPresenceServiceImpl.addUserToBook(user, library.getId(), book.getId());
        });

        verify(bookPresenceRepository, times(1))
                .findAllByLibraryIdAndBookIdAndAvailability(1L, 1L, Availability.AVAILABLE);
        verify(bookPresenceRepository, never()).save(any(BookPresence.class));
        verify(journalService, never()).createJournal(any(Journal.class));
    }

    @Test
    public void removeUserFromBook() {
        when(bookPresenceRepository.findAllByLibraryIdAndBookIdAndUser(library.getId(), book.getId(), user))
                .thenReturn(Optional.of(bookPresence));
        when(journalService.findByBookPresenceIdAndUserIdAndDateOfReturningIsNull(bookPresence.getId(), user.getId()))
                .thenReturn(journal);
        when(journalService.updateJournal(journal.getId(), journal)).thenReturn(journal);
        when(reservationRepository.findFirstByBookIdAndLibraryIdOrLibraryIsNull(book.getId(), library.getId()))
                .thenReturn(reservation);
        when(bookPresenceRepository.save(bookPresence)).thenReturn(bookPresence);

        assertEquals(bookPresence, bookPresenceServiceImpl.removeUserFromBook(user, library.getId(), book.getId()));

        verify(bookPresenceRepository, times(1))
                .findAllByLibraryIdAndBookIdAndUser(library.getId(), book.getId(), user);
        verify(journalService, times(1))
                .findByBookPresenceIdAndUserIdAndDateOfReturningIsNull(bookPresence.getId(), user.getId());
        verify(journalService, times(1)).updateJournal(journal.getId(), journal);
        verify(bookPresenceRepository, times(1)).save(bookPresence);
    }

    @Test
    public void removeUserFromBook_BookPresenceNotFound() {
        when(bookPresenceRepository.findAllByLibraryIdAndBookIdAndUser(library.getId(), book.getId(), user))
                .thenReturn(Optional.empty());

        assertThrows(EntityNotFoundException.class,
                () -> {bookPresenceServiceImpl.removeUserFromBook(user, library.getId(), book.getId());});

        verify(bookPresenceRepository, times(1))
                .findAllByLibraryIdAndBookIdAndUser(library.getId(), book.getId(), user);
        verify(journalService, never()).findByBookPresenceIdAndUserIdAndDateOfReturningIsNull(anyLong(), anyLong());
        verify(journalService, never()).updateJournal(anyLong(), any(Journal.class));
        verify(bookPresenceRepository, never()).save(any(BookPresence.class));
    }

    @Test
    public void addBookToLibrary() {
        when(bookRepository.findById(book.getId())).thenReturn(Optional.of(book));
        when(libraryRepository.findById(book.getId())).thenReturn(Optional.of(library));
        when(bookPresenceRepository.save(any(BookPresence.class))).thenReturn(bookPresence);

        assertEquals(bookPresence, bookPresenceServiceImpl.addBookToLibrary(library.getId(), book.getId()));
        verify(bookRepository, times(1)).findById(book.getId());
        verify(libraryRepository, times(1)).findById(library.getId());
        verify(bookPresenceRepository, times(1)).save(any(BookPresence.class));
    }

    @Test
    public void getByBookId() {
        when(bookPresenceRepository.findAllByBookId(book.getId())).thenReturn(List.of(bookPresence));

        assertEquals(List.of(bookPresence), bookPresenceServiceImpl.getByBookId(book.getId()));
        verify(bookPresenceRepository, times(1)).findAllByBookId(book.getId());
    }

    @Test
    public void getByLibraryId() {
        when(bookPresenceRepository.findAllByLibraryId(library.getId())).thenReturn(List.of(bookPresence));

        assertEquals(List.of(bookPresence), bookPresenceServiceImpl.getByLibraryId(library.getId()));
        verify(bookPresenceRepository, times(1)).findAllByLibraryId(library.getId());
    }

    @Test
    public void getAllBookByLibraryIdAndBookId(){
        when(bookPresenceRepository
                .findAllByLibraryIdAndBookId(library.getId(), book.getId())).thenReturn(List.of(bookPresence));

        assertEquals(List.of(bookPresence),
                bookPresenceServiceImpl.getAllBookByLibraryIdAndBookId(library.getId(), book.getId()));
        verify(bookPresenceRepository, times(1))
                .findAllByLibraryIdAndBookId(library.getId(), book.getId());
    }

    @Test
    public void getAllBookByLibraryIdAndAvailability(){
        when(bookPresenceRepository
                .findAllByLibraryIdAndAvailability(library.getId(), Availability.AVAILABLE)).thenReturn(List.of(bookPresence));

        assertEquals(List.of(bookPresence),
                bookPresenceServiceImpl.getAllBookByLibraryIdAndAvailability(library.getId(), Availability.AVAILABLE));
        verify(bookPresenceRepository, times(1))
                .findAllByLibraryIdAndAvailability(library.getId(), Availability.AVAILABLE);
    }

    @Test
    public void findAllByLibraryIdAndBookIdAndAvailability(){
        when(bookPresenceRepository
                .findAllByLibraryIdAndBookIdAndAvailability(library.getId(), book.getId(), Availability.AVAILABLE))
                .thenReturn(List.of(bookPresence));

        assertEquals(List.of(bookPresence), bookPresenceServiceImpl
                .findAllByLibraryIdAndBookIdAndAvailability(library.getId(), book.getId(), Availability.AVAILABLE));
        verify(bookPresenceRepository, times(1))
                .findAllByLibraryIdAndBookIdAndAvailability(library.getId(), book.getId(), Availability.AVAILABLE);
    }

    @Test
    public void deleteBookPresenceByIdAndLibraryId(){
        doNothing().when(bookPresenceRepository).deleteBookPresenceByIdAndLibraryId(book.getId(), library.getId());

        bookPresenceServiceImpl.deleteBookPresenceByIdAndLibraryId(library.getId(), bookPresence.getId());
        verify(bookPresenceRepository, times(1))
                .deleteBookPresenceByIdAndLibraryId(book.getId(), library.getId());
    }

    @Test
    public void deleteBookPresenceById(){
        doNothing().when(bookPresenceRepository).deleteById(bookPresence.getId());

        bookPresenceServiceImpl.deleteBookPresenceById(bookPresence.getId());
        verify(bookPresenceRepository, times(1)).deleteById(bookPresence.getId());
    }
}

File: src/test/java/com/example/librarymanagement/service/impl/BookPresenceImplTest.kt
package com.example.librarymanagement.service.impl

import com.example.librarymanagement.data.BookDataFactory
import com.example.librarymanagement.data.BookPresenceDataFactory
import com.example.librarymanagement.data.JournalDataFactory
import com.example.librarymanagement.data.LibraryDataFactory
import com.example.librarymanagement.data.ReservationDataFactory
import com.example.librarymanagement.data.UserDataFactory
import com.example.librarymanagement.model.domain.BookPresence
import com.example.librarymanagement.model.enums.Availability
import com.example.librarymanagement.repository.BookPresenceRepository
import com.example.librarymanagement.repository.ReservationRepository
import com.example.librarymanagement.service.BookService
import com.example.librarymanagement.service.JournalService
import com.example.librarymanagement.service.LibraryService
import com.example.librarymanagement.service.UserService
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class BookPresenceImplTest {
    private val bookPresenceRepository: BookPresenceRepository = mockk()
    private val journalService: JournalService = mockk()
    private val libraryService: LibraryService = mockk()
    private val reservationRepository: ReservationRepository = mockk()
    private val bookService: BookService = mockk()
    private val userService: UserService = mockk()
    private val bookPresenceService = BookPresenceServiceImpl(
        bookPresenceRepository,
        journalService,
        reservationRepository,
        bookService,
        libraryService,
        userService
    )
    private val id = "1"

    private val bookPresence = BookPresenceDataFactory.createBookPresence(id)
    private val book = BookDataFactory.createBook(id)
    private val user = UserDataFactory.createUser(id)
    private val journal = JournalDataFactory.createJournal(id)
    private val library = LibraryDataFactory.createLibrary(id)
    private val reservation = ReservationDataFactory.createReservation(id)

    @Test
    fun shouldAddUserToBookWithReservation() {
        //GIVEN
        val reservationTest = reservation.copy(userId = id)
        val expected = listOf(journal)

        every { userService.getUserById(id) } returns user
        every { libraryService.getLibraryById(id) } returns library
        every { reservationRepository.findFirstByBookIdAndLibraryIdOrLibraryIsNull(id, id) }
            .returns(reservationTest)
        every { reservationRepository.deleteById(id) } returns Unit
        every { bookPresenceRepository.addBookToUser(user, id, id) } returns bookPresence
        every { journalService.getJournalByUserId(id) } returns listOf(journal)

        //WHEN
        val actual = bookPresenceService.addUserToBook(id, id, id)

        //THEN
        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) { reservationRepository.deleteById(id) }
        verify(exactly = 1) { bookPresenceRepository.addBookToUser(user, id, id) }
        verify(exactly = 1) { journalService.getJournalByUserId(id) }
    }

    @Test
    fun shouldAddUserToBookWithoutReservation() {
        //GIVEN
        val expected = listOf(journal)

        every { userService.getUserById(id) } returns user
        every { libraryService.getLibraryById(id) } returns library
        every { reservationRepository.findFirstByBookIdAndLibraryIdOrLibraryIsNull(id, id) }
            .returns(null)
        every { bookPresenceRepository.addBookToUser(user, id, id) } returns bookPresence
        every { journalService.getJournalByUserId(id) } returns listOf(journal)

        //WHEN
        val actual = bookPresenceService.addUserToBook(id, id, id)

        //THEN
        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) { userService.getUserById(id) }
        verify(exactly = 1) { libraryService.getLibraryById(id) }
        verify(exactly = 1) { reservationRepository.findFirstByBookIdAndLibraryIdOrLibraryIsNull(id, id) }
        verify(exactly = 1) { bookPresenceRepository.addBookToUser(user, id, id) }
        verify(exactly = 1) { journalService.getJournalByUserId(id) }
    }

    @Test
    fun shouldAddBookToLibrary() {
        //GIVEN
        val bookWithout = book
        val libraryTest = library
        val expected = BookPresence(bookId = id, libraryId = id)

        every { bookService.getBookById(id) } returns bookWithout
        every { libraryService.getLibraryById(id) } returns libraryTest
        every { bookPresenceRepository.save(expected) } returns expected

        //WHEN
        val actual = bookPresenceService.addBookToLibrary(id, id)

        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) { bookService.getBookById(id) }
        verify(exactly = 1) { libraryService.getLibraryById(id) }
        verify(exactly = 1) { bookPresenceRepository.save(expected) }
    }

    @Test
    fun shouldRemoveUserFromBookWithReservation() {
        //GIVEN
        val expected = listOf(journal)

        every { userService.getUserById(id) } returns user
        every { libraryService.getLibraryById(id) } returns library
        every { bookPresenceRepository.removeBookFromUser(user, id, id) } returns bookPresence
        every { journalService.getJournalByUserId(id) } returns listOf(journal)

        //WHEN
        val actual = bookPresenceService.removeUserFromBook(id, id, id)

        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun shouldGetBookPresenceByLibraryId() {
        //GIVEN
        val excepted = listOf(bookPresence)

        every { bookPresenceRepository.findAllByLibraryId(id) } returns listOf(bookPresence)

        //WHEN
        val actual = bookPresenceService.getAllByLibraryId(id)

        //THEN
        Assertions.assertEquals(excepted, actual)
        verify(exactly = 1) { bookPresenceRepository.findAllByLibraryId(id) }
    }

    @Test
    fun shouldGetBookPresenceByBookId() {
        //GIVEN
        val excepted = listOf(bookPresence)

        every { bookPresenceRepository.findAllByBookId(id) } returns listOf(bookPresence)

        //WHEN
        val actual = bookPresenceService.getAllByBookId(id)

        Assertions.assertEquals(excepted, actual)
        verify(exactly = 1) { bookPresenceRepository.findAllByBookId(id) }
    }

    @Test
    fun shouldGetBookPresenceByLibraryIdAndBookId() {
        //GIVEN
        val expected = listOf(bookPresence)

        every { bookPresenceRepository.findAllByLibraryIdAndBookId(id, id) } returns listOf(
            bookPresence
        )

        //WHEN
        val actual = bookPresenceService.getAllBookPresencesByLibraryIdAndBookId(id, id)

        //THEN
        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) { bookPresenceRepository.findAllByLibraryIdAndBookId(id, id) }
    }

    @Test
    fun shouldGetBookPresenceByLibraryIdAndAvailability() {
        //GIVEN
        val expected = listOf(bookPresence)

        every { bookPresenceRepository.findAllByLibraryIdAndAvailability(id, Availability.AVAILABLE) }
            .returns(listOf(bookPresence))

        //WHEN
        val actual = bookPresenceService.getAllBookPresencesByLibraryIdAndAvailability(id, Availability.AVAILABLE)

        //THEN
        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) {
            bookPresenceRepository.findAllByLibraryIdAndAvailability(id, Availability.AVAILABLE)
        }
    }

    @Test
    fun shouldDeleteBookPresenceById() {
        //GIVEN
        every { bookPresenceRepository.deleteById(bookPresence.id!!) } returns Unit

        //WHEN
        bookPresenceService.deleteBookPresenceById(bookPresence.id!!)

        //THEN
        verify(exactly = 1) { bookPresenceRepository.deleteById(bookPresence.id!!) }
    }
}
